Utilities panel + LOD creation

luketimothyjones opened the pull request:
NOTE: Relies on #25 

Automatically generates 3 LOD levels (on top of original mesh) for use in games. LOD levels are as lossless as possible, as creating a proper lossy LOD for the last LOD level is not _easily_ doable in Blender as far as I know. Tested and working in Unity.

Resulting object structure:
![Resulting object structure](https://user-images.githubusercontent.com/6316047/54394382-ecced880-4669-11e9-9d3d-633fae78376d.png)

luketimothyjones commented:
Thought of a potentially better and faster way to do this than using limited dissolve. You had mentioned a long time ago adjusting the curve resolution, but it hadn't clicked in my head how widespread that change would be. Additionally, playing around with the decimate modifier may yield interesting results.

This is good enough for now, but in the near future I'll see if I can make it better.

friggog commented:
Great idea to include this, hadn't crossed my mind despite being included in the original paper. Would recommend glancing over [section 5](https://www2.cs.duke.edu/courses/cps124/spring08/assign/07_papers/p119-weber.pdf) - a lot of it is probably irrelevant, but may be of interest. Yeah also changing the input parameters and regenerating could be an option (as you say curve-res could be an option, but would probably also change the shape of the tree too much). 

We should probably also consider leaves when it comes to LoD as the leaf mesh is pretty crazy in terms of complexity. Again not really sure how to fix this - maybe just regenerating with larger leaf scale and lower leaf number? Is there some operation to create some approximate blended mesh from lots of smaller meshes or something?

luketimothyjones commented:
I have a working version that creates copies of the original curve, lowering the resolution each time and then decimating them instead of using limited dissolve. This strategy works MUCH better and had very little effect on the shape of the tree (though admittedly I've only tested it with our good friend quaking aspen thus far).

luketimothyjones commented:
Pushed the new decimate method

luketimothyjones commented:
I moved the checkboxes down into the utilities section. Not ready for merge; for some reason I'm running into issues with running LOD creation automatically, still working on debugging it.

luketimothyjones commented:
Speaking of debugging, memory access errors make me want to die. Having a hell of a time getting this bug squashed.

luketimothyjones commented:
Woohoo! After spending far too many hours debugging, I have fixed the issue.

As was previously known, `bpy` is not thread-safe, but we've gotten away with using it anyway... until now. Applying modifiers causes some seriously bizarre un-Googleable issues ranging from `RuntimeError`s stopping LOD generation to memory access violations causing hard crashes.

I fixed this by adding a way to queue tasks for execution on the main thread from another thread via a modal operator. UX doesn't change, as both mesh conversion and LOD generation freeze up the GUI when run manually anyway.

Once you've reviewed it this PR is ready for merge 🎉

luketimothyjones commented:
![image](https://user-images.githubusercontent.com/6316047/54713130-217bdd80-4b0b-11e9-8aaa-84071c382c53.png)

friggog commented:
Will test this out later today hopefully, or in the next couple of days if not - does that mean #25 is obsolete then? If so you can close it without merge.

luketimothyjones commented:
Yep, should be.

friggog commented on ch_trees/gui.py:
```diff
@@ -170,27 +167,49 @@ class TreeGen(bpy.types.Operator):
     _scene.custom_tree_load_params_input = _props.EnumProperty(name="", default="ch_trees.parametric.tree_params.quaking_aspen", items=parametric_items)
 
+    # Render inputs; auto-fill path input with user's home directory
+    _scene.render_input = _props.BoolProperty(name="Render", default=False)
```
I feel like this should be "Render after generation" or something to make clear it's automatic.

friggog commented on ch_trees/gui.py:
```diff
@@ -170,27 +167,49 @@ class TreeGen(bpy.types.Operator):
+    _scene.render_output_path_input = _props.StringProperty(name="", default=render_output_path)
+
+    # Convert selected tree to mesh
+    _scene.tree_gen_convert_to_mesh_input = _props.BoolProperty(name="Convert to Mesh", default=False, description="After generation, automatically convert to mesh.")
```
Similarly this should be made clear it is automatic and distinct from the button which currently says the same thing

friggog replied:
Should render also have a button like the other utilities to do it on demand?

friggog commented on ch_trees/gui.py:
```diff
@@ -170,27 +167,49 @@ class TreeGen(bpy.types.Operator):
+    _scene.tree_gen_convert_to_mesh_input = _props.BoolProperty(name="Convert to Mesh", default=False, description="After generation, automatically convert to mesh.")
+
+    # Create LODs
+    _scene.tree_gen_create_lods_input = _props.BoolProperty(name="Create LODs", default=False, description="After generation, create three copies of the tree (meshes) of decreasing quality. The original tree curve will be preserved, but can be converted using the 'Convert To Mesh' button.")
```
As above

luketimothyjones commented:
I'll update the descriptions and add a button for manual rendering
hopefully this morning.

Leaf LOD generation is going to be tricky, as the leaves are all one big
mesh.

I have a local branch in which I've added a "vertex" leaf that creates a
single vertex at the root of the leaf location instead of adding an entire
leaf. I figure it may be possible to use it for generating leaves after the
base tree has been generated. It's not a feature I intend to work on
immediately, but I wanted to see how easy it would be to add the vertex
portion of the implementation.

luketimothyjones commented:
![image](https://user-images.githubusercontent.com/6316047/54967216-158d8280-4f34-11e9-985d-9ba5a62c4756.png)

Look good?

friggog commented:
Looks good, just tested and everything seems to be fine - happy to merge if you are?

luketimothyjones commented:
I'd like to add a parameter set for both curve resolution and decimation
amount, as needs will vary from tree to tree, but I'm not sure when I'll be
able to get that done.

With that said, I think it's good for merge now. I'll get the utilities
documentation done as soon as I can. After that I'll focus on improvements
to the LOD system.
